src/app/editor/code/editor-container/codegen/xamarin-forms/codegen/codegenvisitor.service.ts

```diff
@@ -0,0 +1,109 @@
+import { Template } from '../../shared-codegen.service';
+
+/**
+ * Visitor-like pattern used for code generation purposes, by iterating through an AST and
+ * delegating the codegen to its subclasses.
+ */
+export abstract class CodeGenVisitor {
+    abstract readonly SupportedTemplate: Template;
```

Comment:
This is a relic of the past, it should not be here.

Reply:
Before we started working on the feature, I actually started playing a bit with the codebase, and was trying to remove the giant `switch` statement in the [CodeGenService.generate](/xlayers/xlayers/blob/master/src/app/editor/code/editor-container/codegen/codegen.service.ts#L124) method. The property is a relic of my test, my bad for letting it slip through until the PR.

Reply:
@t-dambacher @LopezArnaud Not a problem, feel free to explore possibilities. We are also investigating ways to improve the structure of CodeGen ヽ(ヅ)ノ

Reply:
@manekinekko @Shikanime Is there a specific place where you listed your current ideas about this particular subject ? Now that we have implemented the codegen for a new language, we have a bit of understanding about the codegen codebase, and have some insights about its strength (genericity, separation between styling and content, decoupled from HMI, first-class support for front-end languages, etc.), current weaknesses (high coupling between the AST navigation and the codegen, the `CodeGenService` class that references each language-specific services where it could benefit from more genericity), and some pain point that are hard to address (relative positioning, shapes and elements recognition). Once we are done with this pull request, we would love to help more where we could. How could we proceed ?

Reply:
We generally share our ideas about Slack or about opening a Github as an RFC. If it is large enough to require special attention, we describe the idea in more detail in a design document.

Here are some example:

- Material Angular: https://github.com/angular/material2/wiki/Design-doc-directory
- xLayers Flutter implementation: https://docs.google.com/document/d/15dige0j42VoIkUMq8AETkJWYYoDnN0WPZRsQAHbOffQ/edit?usp=sharing

After looking at the Xamarin CodeGen I found some similarities and some generic services that can lead to a shared service such as the `XML` service with `openTag` and `closeTag`. And  ultimately used in a higher level service like a `Web` service with `generateTemplate`.
Which are quite similar to the current `SharedCodegen`.

Current structure:

![Xlayers design-Page-4 (2)](https://user-images.githubusercontent.com/22115108/56644157-6ef3e900-667b-11e9-9905-6014a2892ac5.jpg)

Ideal futur structure:

![Xlayers design-Page-3 (1)](https://user-images.githubusercontent.com/22115108/56644160-70251600-667b-11e9-9d62-4b309400495f.jpg)

Reply:
Thanks for the great explanation. I will look at all this ASAP.